Load the sequence

In [3]:
with open("../output/combined_text.tex", "r", encoding="utf-8") as f:
    text_sequence = f.read()
len(text_sequence)

12379

BPE algorithm: byte pair encoding tokenizaton. for encoding string of text into smaller strings by creating and using a translation table. 

I am using the minbpe repo to tokenize the aquence of text. 

In [4]:
import sys
sys.path.append('..')

Start by training the tokenizer on the text sequence that you saved in the previous notebook.

In [9]:
from minbpe import BasicTokenizer
tokenizer = BasicTokenizer()
tokenizer.train(text_sequence, vocab_size=1024)

Visualize the vocabulary.

In [10]:
vocab = tokenizer.vocab
vocab_size = len(vocab)
print(f"Vocabulary size: {vocab_size}")

Vocabulary size: 1024


Test the tokenizer.

In [ ]:
tokenizer.encode("Arun Ayush")


[65, 114, 289, 32, 65, 121, 334, 104]

In [13]:
tokenizer.decode([65, 114, 289, 32, 65, 121, 334, 104])

'Arun Ayush'

Add special tokens to the vocabulary. These tokens are going to be used a lot in the fine-tuning step.

In [14]:
max_vocab_id = list(tokenizer.vocab.keys())[-1]
tokenizer.special_tokens = {
    "<|startoftext|>": max_vocab_id + 1,
    "<|separator|>": max_vocab_id + 2,
    "<|endoftext|>": max_vocab_id + 3,
    "<|unk|>": max_vocab_id + 4
}

tokenizer.special_tokens = {
    "<|startoftext|>": max_vocab_id + 1,
    "<|separator|>": max_vocab_id + 2,
    "<|endoftext|>": max_vocab_id + 3,
    "<|unk|>": max_vocab_id + 4
}
This is defining special-use tokens with IDs that won't conflict with existing tokens in the vocab.

Finding the max ID in the current vocab.

Creating 4 custom tokens after that ID.

Ensuring they don’t conflict with existing tokens.

In [15]:
len(tokenizer.encode(text_sequence))

4113

I have more than 4113 tokens for training and validation. This is pretty good for our initial point, but if you can add more, that would be even better.

In [17]:
import os

os.makedirs("../output/tokenizer", exist_ok=True)
tokenizer.save(file_prefix="../output/tokenizer/my_tokenizer")
